### DISCUSSION TITLE
Don't require primary key on M2M junction table

### DISCUSSION THREAD

**Discussion by AaronBeaudoin** [directus/directus#11897] [Feature Requests]
Hi. I've been experimenting with getting Directus up and running in an existing PostgreSQL database managed by a fairly popular open-source ERP system called [Odoo](https://github.com/odoo/odoo).

An issue that I'm running into is that Odoo's ORM creates junction tables for M2M relationships **without** a primary key. The junction table has only the 2 necessary columns, one to reference the "parent" table's ID and one to reference the "child" table's ID. I assume this is for efficiency. As far as I can tell Odoo also [doesn't provide an option](https://www.odoo.com/documentation/13.0/developer/reference/addons/orm.html#odoo.fields.Many2many) to create a primary key on junction tables.

As a result, the table is broken in Directus. I get an error like the following in the console:

```
Collection "<junction_table_name>" doesn't have a primary key column and will be ignored
```

The best workaround that I might be able to achieve with some trickery is to potentially configure Odoo's ORM to let me manually create the junction table and then create it manually with a primary key to make Directus happy.

However, this honestly seems like overkill, an unnecessary hack to get around a limitation in Directus. I think a **better solution** would be to simply not require a primary key for junction tables in Directus. What are your thoughts?

**Comment by rijkvanzanten** [comment]
Right now, there's no difference between "junction" collections and regular collections in Directus at all. In order for Directus to be able to manage the rows contained in the collection, there needs to be a way to uniquely identify the rows, which is currently done through the reliance of a primary key column. Once we support compound primary keys, we can start supporting junction collections without a separate primary key column, as the unique combination of "parent" and "child" would be considered the identifier. 👍🏻

**Reply by AaronBeaudoin** [reply]
Are compound primary keys planned on the roadmap or is it just a known limitation for the foreseeable future? Also, would there need to be a compound primary key at the database level or could it be a "calculated" thing in Directus?

**Reply by rijkvanzanten** [reply]
> Are compound primary keys planned on the roadmap or is it just a known limitation for the foreseeable future?

Yes it's definitely something we're interested in officially supporting, but not something that's actively being developed right now. 

> Also, would there need to be a compound primary key at the database level or could it be a "calculated" thing in Directus?

We would most likely rely on the database indexes in that case as well, just like for regular primary keys and other indexes/constraints 👍🏻

**Reply by sxsws** [reply]
Am I right in assuming that nuxt is the orm used? Does that enforce the active record pattern?

**Reply by xaviNo4** [reply]
hi, Aaron, how about your work on odoo with directus? I am going to do the similar work.

**Comment by frdrk** [comment]
I too would be very much interested in a solution to this limitation.

**Reply by rijkvanzanten** [reply]
Directus would have to have some sort of way to identify what row is being used though. When you're looking to have a junction table without a primary key, would you be using the both junction fields as the composite primary key?

**Reply by frdrk** [reply]
Yes I believe that's how composite keys work. But I'm by no means an SQL expert so I might be wrong though.

**Reply by sxsws** [reply]
yes, that's right. the two keys make up one primary key and are unique, e.g. there can only be one record where a Director:14 has a relation to Film:20. It would be wrong for another record to exist as, e.g. you can't have two records to the same film and director twice. Director:14 has a relation to Film:30 would be another relation for Director14. 

So, if you wanted to get all films that the director was involved in, you'd just search the table for where DirectorId is whatever, and if you want to know the directors for a film, you'd search by FilmId

This allows you to add additional columns to the junction table for additional data concerning the relation, in this example, maybe dfirector's salary, whether they were the 'lead' director, etc.

Whether nuxt allows this is another story. It won't if it enforces the Active Record pattern.

**Reply by rijkvanzanten** [reply]
> This allows you to add additional columns to the junction table for additional data concerning the relation, in this example, maybe dfirector's salary, whether they were the 'lead' director, etc.

This is technically already possible as well 👍🏻 

The main question to answer is how to access rows with composite primary keys from a URL. We'll have to come up with some sort of separator character and saved order in order to access the row. Alternatively, we could add some sort of column identifier into the mix, like:

```
// Current, single primary key:
/items/junction/15

// Composite?
/items/junction/15:31
/items/junction/film:15|director:31
```

(random example, don't think `:` is a valid character in a URL path)

**Reply by sxsws** [reply]
From [this](https://stackoverflow.com/questions/1547899/which-characters-make-a-url-invalid) SO post, the following are valid characters:
`ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~:/?#[]@!$&'()*+,;=`

so both `:` and `,` are valid but `|` is not. Seeing `[]` as being valid may be good to use to denote the list-ness of the values

```
/items/junction/film:15,director:31

// array-like syntax
items/junction/[31:15]  // key values defined in alphabetical order (director:film) 
items/junction/[31,15]  // my preference for short-hand format
items/junction/[film:15,director:31]  // keys can be defined in any order and my overall preference
```
Maybe both short-hand and full declaration can be supported, but the url with identifiers is more self-documenting and easier to remember in six-months time when you have to change something instead of having to remember which column is first alphabetically!

**Reply by rijkvanzanten** [reply]
> From [this](https://stackoverflow.com/questions/1547899/which-characters-make-a-url-invalid) SO post, the following are valid characters: `ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~:/?#[]@!$&'()*+,;=`

I _think_ some of those characters are only valid on some parts of the URL though, not necessarily as part of the path. For example, every character after a `?` is considered the query string, not part of the path. Similarly, I believe `@` is only allowed as part of the authentication part of the resource locator (like `username:password@directus.io/path/to/resource?query=value`) but I might be mistaken. 


> Maybe both short-hand and full declaration can be supported, but the url with identifiers is more self-documenting and easier to remember in six-months time when you have to change something instead of having to remember which column is first alphabetically!

I'm wondering what order would be the expected order for a shorthand 🤔 Would it be alphabetical? Should it be based on the order configured in the schema? Should it be based on the ordinal order of the columns in the table? Maybe based on the location in the configure key? (IIRC you can set the primary key to `(colA, colB)` or `(colB, colA)` 🤔

**Comment by sxsws** [comment]
It would have to be alphanumerical because the order of columns can be ALTERed, That way, it prevents all the problems that are foreseeable for the other possibilities if the table is modified externally..

```
ALTER TABLE `table_name`
    MODIFY COLUMN `column_name1` varchar(100) AFTER `column_name2`;
```

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Reply by glowkeeper** [reply]
I'm guessing that because I'm seeing _WARN Collection "myCompoundKeyCollection" doesn't have a primary key column and will be ignored_ on a junction table with a compound key, Directus still doesn't support those, which is a shame 😞

**Reply by glowkeeper** [reply]
Without that support, what are the workarounds? I can think of one or two, but I'd be interested to hear what people do...

**Reply by glowkeeper** [reply]
Any chance you could reopen this @rijkvanzanten? After all, it's still an issue...

**Reply by rijkvanzanten** [reply]
> Any chance you could reopen this @rijkvanzanten? After all, it's still an issue...

We've never supported compound/composite primary/foreign keys, nor claim to do so, so we're treating this as a new feature rather a bugfix 👍🏻  Feel free to open a new feature request detailing your exact requirements, and if [there's enough want/need from the larger community](https://docs.directus.io/contributing/feature-request-process.html), we can add support to core 🙂
